### PR TITLE
Init

### PR DESCRIPTION
**Pull request by smiller171**

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -1,16 +1,135 @@
...
+
+resource "aws_secretsmanager_secret" "pubkey" {
+  name_prefix = "${var.name}-pubkey-"
+  description = var.secret_description
+  tags = merge(
+    var.tags,
+    {
+      Name = "${var.name}-pubkey"
+    }
+  )
```
Comment (sblack4) [inline main.tf:115]: Don't we want to store the public key as an EC2 key pair? https://www.terraform.io/docs/providers/aws/r/key_pair.html

Comment (smiller171) [reply main.tf:115]: I think it's best to do that in the parent module if desired. That's not necessarily why you're creating a keypair and you can't pull the pubkey from an ec2_key_pair as far as I can tell.

This isn't actually secret, but a secret is an easy way to manage it.

I did add a commit to add the pubkey as an output though.

**Review by sblack4 — APPROVED** [review]
:shipit:

File: lambda/index.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
+const AWS = require('aws-sdk')
+const forge = require('node-forge')
+
+const rsa = forge.pki.rsa
+
+const secretsmanager = new AWS.SecretsManager()
+
+exports.handler = async (event, context) => {
+  try {
+    const keypair = rsa.generateKeyPair({ bits: 2048, e: 0x10001 })
```
Comment (cdaniluk) [inline lambda/index.js]: make the bits configurable? 4096 bit keys are not supported everywhere (fuck you, AWS) but are supported in most things.

File: main.tf [hunk 3]
```diff
@@ -1,16 +1,140 @@
...
+
+    principals {
+      type        = "Service"
+      identifiers = ["lambda.amazonaws.com"]
+    }
+  }
+}
+
+resource "aws_iam_role" "this" {
+  name_prefix = "${var.name}-"
```
Comment (cdaniluk) [inline main.tf:31]:
```suggestion
  name_prefix = "${var.name}"
```
I thought you were pushing us away from the superfluous dash

File: main.tf [hunk 4]
```diff
@@ -1,16 +1,140 @@
...
+}
+
+resource "aws_iam_role_policy_attachment" "lambda_basic_execution" {
+  role       = aws_iam_role.this.name
+  policy_arn = "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
+}
+
+resource "aws_iam_role_policy_attachment" "secretsmanager_read_write" {
+  role       = aws_iam_role.this.name
+  policy_arn = "arn:aws:iam::aws:policy/SecretsManagerReadWrite"
```
Comment (cdaniluk) [inline main.tf]: this is unnecessarily broad

File: main.tf [hunk 5]
```diff
@@ -1,16 +1,140 @@
...
+  tags = merge(
+    var.tags,
+    {
+      "Name" = var.name
+    }
+  )
+}
+
+resource "aws_secretsmanager_secret" "privkey" {
+  name_prefix = "${var.name}-privkey-"
```
Comment (cdaniluk) [inline main.tf:97]:
```suggestion
  name_prefix = "${var.name}-privkey"
```

File: main.tf [hunk 6]
```diff
@@ -1,16 +1,140 @@
...
+  tags = merge(
+    var.tags,
+    {
+      Name = "${var.name}-privkey"
+    }
+  )
+}
+
+resource "aws_secretsmanager_secret" "pubkey" {
+  name_prefix = "${var.name}-pubkey-"
```
Comment (cdaniluk) [inline main.tf:108]:
```suggestion
  name_prefix = "${var.name}-pubkey"
```

File: variables.tf [hunk 7]
```diff
@@ -4,8 +4,20 @@ variable "name" {
...
+  type        = string
+}
+
 variable "tags" {
   default     = {}
   description = "User-Defined tags"
   type        = map(string)
 }
+
+variable "keepers" {
```
Comment (cdaniluk) [inline variables.tf]: sort

File: variables.tf [hunk 8]
```diff
@@ -4,8 +4,20 @@ variable "name" {
   type        = string
 }
 
+variable "secret_description" {
+  default     = "A password created by Terraform" #tfsec:ignore:GEN001
```
Comment (cdaniluk) [inline variables.tf]:
```suggestion
  default     = "An SSH key secret by Terraform" #tfsec:ignore:GEN001
```

**Review by cdaniluk — CHANGES_REQUESTED** [review]

**Comment by cdaniluk** [comment]
checks need fixed too..

File: outputs.tf [hunk 9]
```diff
@@ -1,5 +1,24 @@
+output "result" {
```
Comment (cdaniluk) [inline outputs.tf:16]: You gotta export the secret ARN. Pretty important to the utility of the module.

File: outputs.tf [hunk 10]
```diff
@@ -1,5 +1,24 @@
...
+  description = "String result of Lambda execution"
+  value       = module.lambda_invocation_result.stdout
+}
+
+output "invocation_stdout" {
```
Comment (cdaniluk) [inline outputs.tf:21]: Are these just in for debugging purposes? If so maybe nix? If they are worth keeping, that's fine. Seems weird though. Maybe a TODO explaining?

File: outputs.tf (hunk 9 above)
Comment (cdaniluk) [reply outputs.tf:16]: Maybe the stdout is needed because of the dependency, but the others?

File: main.tf [hunk 11]
```diff
@@ -1,16 +1,140 @@
...
+  source_code_hash = try(filebase64sha256(local.zipfile), null)
+
+  lifecycle {
+    ignore_changes = [
+      filename,
+      last_modified
+    ]
+  }
+
+  tags = merge(
```
Comment (cdaniluk) [inline main.tf:84]: tags var goes above lifecycle block

File: main.tf [hunk 12]
```diff
@@ -1,16 +1,140 @@
...
+  policy_arn = "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
+}
+
+resource "aws_iam_role_policy_attachment" "secretsmanager_read_write" {
+  role       = aws_iam_role.this.name
+  policy_arn = "arn:aws:iam::aws:policy/SecretsManagerReadWrite"
+}
+
+resource "aws_lambda_function" "this" {
+  depends_on = [
```
Comment (cdaniluk) [inline main.tf]: depends on at the bottom

File: main.tf (hunk 4 above)
Comment (cdaniluk) [reply main.tf]: TF creates the secret, so the Lambda can just be given access to PutSecretVersion against that.

**Review by cdaniluk — CHANGES_REQUESTED** [review]

File: main.tf (hunk 3 above)
Comment (smiller171) [reply main.tf:31]: Nope. IMO it isn't superfluous. It creates a visual break between the prefix and the terraform-generated stuff.

File: outputs.tf (hunk 10 above)
Comment (smiller171) [reply outputs.tf:21]: They are attempting to control the dependency flow with some of the modules I'm using. TF 0.13.0 will make this a lot better. I'll put in a comment.

File: lambda/index.js [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+  try {
+    const keypair = rsa.generateKeyPair({ bits, e: 0x10001 })
+    await secretsmanager.putSecretValue({
+      SecretId: event.pubkey_secret_name,
+      SecretString: forge.ssh.publicKeyToOpenSSH(keypair.publicKey)
+    }).promise()
+    await secretsmanager.putSecretValue({
+      SecretId: event.privkey_secret_name,
+      SecretString: forge.ssh.privateKeyToOpenSSH(keypair.privateKey)
+    }).promise()
```
Comment (sblack4) [inline lambda/index.js:20]: Why are we doing this twice?

**Review by sblack4 — APPROVED** [review]

**Pull request closed without merging**